File: Data Structures & Algorithms/count-subsequences/submission-1.cpp
class Solution {
public:
    vector<vector<int>>cache;
    int recurse(string s,string t,int i,int j){
        if(j == t.size()){  
            return 1;
        }

        if(i == s.size()){  
            return 0;
        }

        if(cache[i][j] != -1){
            return cache[i][j];
        }
        
        int ans = 0;
        if(s[i] == t[j]){
            ans += recurse(s,t,i+1,j+1);
        }

        ans += recurse(s,t,i+1,j);

        return cache[i][j] = ans;

    }
    int numDistinct(string s, string t) {
        /*
            c   a   a   a   t
                        |
            c   a   t
                    |
        */
        int m = s.size();
        int n = t.size();
        cache.resize(m,vector<int>(n,-1));
        return recurse(s,t,0,0);
    }
};

File: Data Structures & Algorithms/jump-game-ii/submission-1.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        """
            brute force recursion ofc

            min number of jumps

            2   4   1   1   1   1   

        """

        n = len(nums)
        cache = {}
        
        def dfs(idx: int) -> int:
            if idx >= n-1:
                return 0
            
            if idx in cache:
                return cache[idx]
            
            ans = float("inf")
            for i in range(1,nums[idx]+1):
                ans = min(ans, 1 + dfs(idx + i))

            cache[idx] = ans
            return ans
        
        result = dfs(0)
        return result





